vortex-serde/src/layouts/pruning.rs

```diff
@@ -0,0 +1,538 @@
+            Some(Arc::new(BinaryExpr::new(
+                Arc::new(BinaryExpr::new(
+                    Arc::new(BinaryExpr::new(
+                        min_col.clone(),
+                        Operator::NotEq,
+                        replaced_min.clone(),
+                    )),
+                    Operator::Or,
+                    Arc::new(BinaryExpr::new(
+                        replaced_min,
+                        Operator::NotEq,
+                        max_col.clone(),
+                    )),
+                )),
+                Operator::And,
+                Arc::new(BinaryExpr::new(
+                    Arc::new(BinaryExpr::new(
+                        min_col,
+                        Operator::NotEq,
+                        replaced_max.clone(),
+                    )),
+                    Operator::Or,
+                    Arc::new(BinaryExpr::new(replaced_max, Operator::NotEq, max_col)),
+                )),
+            )))
+        }
+        op @ Operator::Gt | op @ Operator::Gte => {
+            let max_col = Arc::new(Column::new(builder.column_stat(Stat::Max)));
+            let replaced_min = builder.rewrite_other_exp(Stat::Min);
+
+            Some(Arc::new(BinaryExpr::new(max_col, op, replaced_min)))
+        }
+        op @ Operator::Lt | op @ Operator::Lte => {
+            let min_col = Arc::new(Column::new(builder.column_stat(Stat::Min)));
+            let replaced_max = builder.rewrite_other_exp(Stat::Max);
+
+            Some(Arc::new(BinaryExpr::new(min_col, op, replaced_max)))
+        }
+        _ => None,
+    };
+    expr.map(|e| (e, builder.into_references()))
+}
+
+fn replace_column_with_stat(
+    expr: &Arc<dyn VortexExpr>,
+    stat: Stat,
+    references: &mut HashMap<Field, ColumnStat>,
+) -> Option<Arc<dyn VortexExpr>> {
+    if let Some(col) = expr.as_any().downcast_ref::<Column>() {
+        let (column, stat_ref) = column_stat_reference(col.field(), stat);
+        references.insert(column.clone(), stat_ref);
+        return Some(Arc::new(Column::new(column)));
+    }
+
+    if let Some(bexp) = expr.as_any().downcast_ref::<BinaryExpr>() {
+        let rewritten_lhs = replace_column_with_stat(bexp.lhs(), stat, references);
+        let rewritten_rhs = replace_column_with_stat(bexp.rhs(), stat, references);
+        if rewritten_lhs.is_none() && rewritten_rhs.is_none() {
+            return None;
+        }
+
+        let lhs = rewritten_lhs.unwrap_or_else(|| bexp.lhs().clone());
+        let rhs = rewritten_rhs.unwrap_or_else(|| bexp.rhs().clone());
+
+        return Some(Arc::new(BinaryExpr::new(lhs, bexp.op(), rhs)));
+    }
+
+    None
+}
+
+fn column_stat_reference(field: &Field, stat: Stat) -> (Field, ColumnStat) {
+    let new_field = stat_column_name(field, stat);
+    (new_field, ColumnStat::new(field.clone(), stat))
+}
+
+fn stat_column_name(field: &Field, stat: Stat) -> Field {
+    match field {
+        Field::Name(n) => Field::Name(format!("{n}_{stat}")),
```

Comment:
if my schema has columns named `a` and `a_min` will this conflict?

Reply:
if we had proper support for nested types this would be as a column in a struct array. Right now it's a bit of a hack since the only client is datafusion and it doesn't support pushdown on nested columns

Reply:
This is next thing to tackle once we have basic support in place